packages/passwordless-auth-construct/src/magic-link/magic_link_challenge_service.ts

```diff
@@ -2,23 +2,166 @@ import {
   CreateAuthChallengeTriggerEvent,
   VerifyAuthChallengeResponseTriggerEvent,
 } from 'aws-lambda';
-import { ChallengeService } from '../types.js';
+import { logger } from '../logger.js';
+
+import { MagicLink, SignedMagicLink } from '../models/magic_link.js';
+import {
+  ChallengeService,
+  MagicLinkConfig,
+  SigningService,
+  StorageService,
+} from '../types.js';
+import { DeliveryServiceFactory } from '../factories/delivery_service_factory.js';
+import { validateDeliveryCodeDetails } from '../common/validate_delivery_code_details.js';
 
 /**
  * Magic Link Challenge Service Implementation.
  */
 export class MagicLinkChallengeService implements ChallengeService {
+  /**
+   * Creates a new MagicLinkChallengeService instance.
+   * @param deliveryServiceFactory - A factory that provides delivery services.
+   * @param config - The config for the service.
+   * @param signingService - The service used to sign links.
+   * @param storageService - the service used to store links.
+   */
+  constructor(
+    private deliveryServiceFactory: DeliveryServiceFactory,
+    private config: MagicLinkConfig,
+    private signingService: SigningService,
+    private storageService: StorageService<SignedMagicLink>
+  ) {}
   public readonly signInMethod = 'MAGIC_LINK';
+
   public createChallenge = async (
     event: CreateAuthChallengeTriggerEvent
   ): Promise<CreateAuthChallengeTriggerEvent> => {
-    // TODO: implement Magic Link
-    return event;
+    logger.info('Starting Create Challenge for Magic Link');
+    // validate redirect URI and delivery details
+    const redirectUri = this.validateRedirectUri(event.request);
+    const { deliveryMedium, destination } = validateDeliveryCodeDetails(event);
+    logger.debug(
+      `Delivery medium: ${deliveryMedium}, destination: ${destination}`
+    );
+    const deliveryService =
+      this.deliveryServiceFactory.getService(deliveryMedium);
+
+    // create magic link
+    logger.info('Creating Magic Link');
+    const userId = event.request.userAttributes.sub;
+    const { userName: username, userPoolId } = event;
+    const { linkDuration } = this.config;
+    const magicLink = MagicLink.create(
+      userPoolId,
+      username,
+      linkDuration?.toSeconds()
```

Comment:
question: Do we expect this to ever be null? Are we handling when this is null gracefully?